### DISCUSSION TITLE
Using Websocket server with Bearer token

### DISCUSSION THREAD

**Discussion by schroe-ja** [actix/actix-web#3506] [Q&A]
Hi, 

I wrote an HTTP(S) endpoint using actix-web which will be upgraded to a websocket connection. I'm also using JWT for authorization. How can I pass the Bearer token as a client? As far as I have read I'm not allowed to add a new header value. Is there something like socket.io that runs in background because socket.io is able to pass the Bearer token.

**Comment by Samielakkad** [comment]
For browser clients, the annoying but important detail is that the standard `WebSocket` constructor does not let you set arbitrary request headers. So this is not really an Actix limitation; it is a browser WebSocket API limitation.

You have a few practical options:

1. Use cookies if the WebSocket is same-site / browser-based

If your normal HTTP login sets an HttpOnly secure cookie, the browser will send that cookie during the WebSocket upgrade request. Then the Actix handler can validate it from the `HttpRequest` before calling `ws::start`:

```rust
async fn ws_route(
    req: HttpRequest,
    stream: web::Payload,
) -> Result<HttpResponse, Error> {
    let cookie = req.cookie(access_token);
    // validate cookie/JWT here before accepting the upgrade

    ws::start(MyWsActor, &req, stream)
}
```

For browser apps, this is usually the cleanest option.

2. Put a short-lived token in the query string

```text
wss://example.com/ws?token=...
```

Then read it from `req.query_string()` or use `web::Query`. If you do this, make the token short-lived and avoid logging full URLs, because query strings often end up in access logs, browser history, proxy logs, etc.

3. Send auth as the first WebSocket message

Accept the connection, require the first message to be an auth message, and close the socket immediately if validation fails. This works, but it means the upgrade itself is accepted before authentication, so you need a strict timeout and no privileged actions before auth succeeds.

4. Non-browser clients can send `Authorization`

If the client is not a browser, many WebSocket clients do allow headers, including `Authorization: Bearer ...`. In that case Actix can read it from `req.headers()` during the upgrade.

I would avoid trying to hide a bearer token in `Sec-WebSocket-Protocol` unless you fully control both sides and understand the tradeoff. That header is meant for subprotocol negotiation, not general auth, and it can make proxies/debugging more confusing.

So the usual recommendation is: cookies for browser apps, explicit `Authorization` headers for non-browser clients, and short-lived query tokens only when cookies are not a good fit.
